substrate/frame/assets-holder/src/impls.rs

```diff
@@ -0,0 +1,253 @@
+		who: &T::AccountId,
+		amount: Self::Balance,
+	) -> Result<Option<Self::Balance>, DispatchError> {
+		pallet_assets::Pallet::<T, I>::write_balance(asset, who, amount)
+	}
+
+	fn set_total_issuance(asset: Self::AssetId, amount: Self::Balance) {
+		pallet_assets::Pallet::<T, I>::set_total_issuance(asset, amount)
+	}
+
+	fn decrease_balance(
+		asset: Self::AssetId,
+		who: &T::AccountId,
+		amount: Self::Balance,
+		precision: Precision,
+		preservation: Preservation,
+		force: Fortitude,
+	) -> Result<Self::Balance, DispatchError> {
+		pallet_assets::Pallet::<T, I>::decrease_balance(
+			asset,
+			who,
+			amount,
+			precision,
+			preservation,
+			force,
+		)
+	}
+
+	fn increase_balance(
+		asset: Self::AssetId,
+		who: &T::AccountId,
+		amount: Self::Balance,
+		precision: Precision,
+	) -> Result<Self::Balance, DispatchError> {
+		pallet_assets::Pallet::<T, I>::increase_balance(asset, who, amount, precision)
+	}
+}
+
+impl<T: Config<I>, I: 'static> UnbalancedHold<T::AccountId> for Pallet<T, I> {
+	fn set_balance_on_hold(
+		asset: Self::AssetId,
+		reason: &Self::Reason,
+		who: &T::AccountId,
+		amount: Self::Balance,
+	) -> DispatchResult {
+		let mut holds = Holds::<T, I>::get(asset.clone(), who);
+		let mut increase = true;
+		let mut delta = amount;
+
+		if let Some(item) = holds.iter_mut().find(|x| &x.id == reason) {
+			delta = item.amount.max(amount) - item.amount.min(amount);
+			increase = amount > item.amount;
+			item.amount = amount;
+			holds.retain(|x| !x.amount.is_zero());
+		} else {
+			if !amount.is_zero() {
+				holds
+					.try_push(IdAmount { id: *reason, amount })
+					.map_err(|_| Error::<T, I>::TooManyHolds)?;
+			}
+		}
+
+		let held_amount = HeldBalances::<T, I>::get(asset.clone(), who).unwrap_or_else(Zero::zero);
+
+		let held_amount = if increase {
+			held_amount.checked_add(&delta).ok_or(ArithmeticError::Overflow)?
+		} else {
+			held_amount.checked_sub(&delta).ok_or(ArithmeticError::Underflow)?
+		};
+
+		if held_amount.is_zero() {
+			HeldBalances::<T, I>::remove(asset.clone(), who);
+		} else {
+			HeldBalances::<T, I>::insert(asset.clone(), who, held_amount);
+		}
+
+		Holds::<T, I>::insert(asset, who, holds);
+
+		Ok(())
+	}
+}
+
+impl<T: Config<I>, I: 'static> MutateHold<T::AccountId> for Pallet<T, I> {
+	fn done_hold(
+		asset_id: Self::AssetId,
+		reason: &Self::Reason,
+		who: &T::AccountId,
+		amount: Self::Balance,
+	) {
+		Self::deposit_event(Event::<T, I>::Held {
+			asset_id,
+			who: who.clone(),
+			reason: *reason,
+			amount,
+		});
+	}
+
+	fn done_release(
+		asset_id: Self::AssetId,
+		reason: &Self::Reason,
+		who: &T::AccountId,
+		amount: Self::Balance,
+	) {
+		Self::deposit_event(Event::<T, I>::Released {
+			asset_id,
+			who: who.clone(),
+			reason: *reason,
+			amount,
+		});
+	}
+
+	fn done_burn_held(
+		asset_id: Self::AssetId,
+		reason: &Self::Reason,
+		who: &T::AccountId,
+		amount: Self::Balance,
+	) {
+		Self::deposit_event(Event::<T, I>::Slashed {
```

Comment:
maybe better to have the event named `Burned` then?

Reply:
I partially agree. IMO there's no reason why this should be called `Slashed`, but there's a slight chance that it might be a preference of keeping this name for historical reasons (wdyt @muharem?)

Reply:
I think it should be `Burned`